### PR TITLE
Implement checking of edge whitespaces

### PR DESCRIPTION
**Pull request by wadimw**

### CODE REVIEW

File: src/rules/ResourceEdgeWhitespace.js [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+            results.push(
+                new Result({
+                    ...resultMetaProps,
+                    description: "Leading whitespace in target does not match leading whitespace in source",
+                    highlight:
+                        `Source: ` +
+                        `<e0>${source.slice(0, whitespaces.source.leading.length)}</e0>` +
+                        source.slice(whitespaces.source.leading.length) +
+                        `\n` +
+                        `Target: ` +
```
Comment (ehoogerbeets) [inline src/rules/ResourceEdgeWhitespace.js:80]: With this sort of highlight in the lint output, the source will be displayed on the screen twice, once from the source field, and here again in the highlight field. The highlight was intended to display just the error string, and perhaps even just a snippet of the string that contained an error if the string happened to be extremely long. I would truncate the source/target to just a few characters.

File: src/rules/ResourceEdgeWhitespace.js [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+                        `<e0>${target.slice(0, whitespaces.target.leading.length)}</e0>` +
```
Comment (ehoogerbeets) [inline src/rules/ResourceEdgeWhitespace.js:81]: The second highlight in the same string should be <e1></e1>

File: test/testRules.js [hunk 3]
```diff
@@ -1183,6 +1184,208 @@ export const testRules = {
...
+                    severity: "error",
+                    pathName: "x/y",
+                    locale: "de-DE",
+                    source: " some source string ",
+                    id: "resource-edge-whitespace.both-spaces-missing",
+                    description: "Trailing whitespace in target does not match trailing whitespace in source",
+                    highlight: "Source:  some source string<e0> </e0>\nTarget: some target string<e0></e0>",
+                }),
+            ],
+        ]);
```
[the remark this answers is not in the document]
Comment (wadimw) [reply test/testRules.js:1442]: Implemented in d710793

File: src/rules/ResourceEdgeWhitespace.js (hunk 1 above)
Comment (wadimw) [reply src/rules/ResourceEdgeWhitespace.js]: Changed in 26588ee as per discussion (truncated, but kept both source and target for positional reference)

File: src/rules/ResourceEdgeWhitespace.js (hunk 2 above)
Comment (wadimw) [reply src/rules/ResourceEdgeWhitespace.js]: Fixed in 26588ee

File: docs/resource-edge-whitespace.md [hunk 4]
```diff
@@ -0,0 +1,16 @@
+# resource-edge-whitespace
+
+Ensure that sequence of whitespaces at either edge of source and target match each other.
+
+If a string extracted for translation from source code has either leading or trailing whitespace, it could mean that it has been used in string concatenation (even though it's incorrect!), so the edge whitespaces might be significant when displaying it in the UI - this rule will detect that such whitespace has been ommited.
```
[the remark this answers is not in the document]
Comment (wadimw) [reply docs/resource-edge-whitespace.md]: Fixed in 8d917d1

**Comment by ehoogerbeets** [comment]
Don't forget to add a note in the release notes section of the README.md that this rule was added

File: test/testRules.js [hunk 5]
```diff
@@ -1183,6 +1184,324 @@ export const testRules = {
...
+                pathName: "x/y",
+                locale: "de-DE",
+                source: " ",
+                id: "resource-edge-whitespace.spaces-only-extra",
+                description: "Leading whitespace in target does not match leading whitespace in source",
+                highlight: `Source: "<e0> </e0>" Target: "<e1>  </e1>"`,
+            })
+        );
         test.done();
     }
```
Comment (ehoogerbeets) [inline test/testRules.js:1506]: What happens if the target string is empty or missing? Does this rule properly ignore it?

Comment (wadimw) [reply test/testRules.js:1506]: This is accounted for in line 50: if either source or target is not a string (so also undefined), checks don't run on that resource

Comment (ehoogerbeets) [reply test/testRules.js:1506]: What I meant is that we should probably have a unit test to verify those things. That way, if someone mucks with the function in the future, we have those guardrails. (We should probably look for code coverage tools for the ilib code.)

**Comment by ehoogerbeets** [comment]
This rule also needs to be added to the list of built-in rules in PluginMgr.js:180 and to the list of rules in the generic ruleset at PluginMgr.js:87

Also, needs a line in the release notes section in the README.md

File: src/rules/ResourceEdgeWhitespace.js [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+        };
+
+        if (whitespaces.target.leading !== whitespaces.source.leading) {
+            results.push(
+                new Result({
+                    ...resultMetaProps,
+                    description: "Leading whitespace in target does not match leading whitespace in source",
+                    highlight:
+                        `Source: ` +
+                        `<e0>${source.slice(0, whitespaces.source.leading.length)}</e0>` +
```
[the remark this answers is not in the document]
Comment (wadimw) [reply src/rules/ResourceEdgeWhitespace.js]: Added in a48653e (depends on https://github.com/iLib-js/i18nlint-common/pull/7)

File: test/testRules.js (hunk 5 above)
Comment (wadimw) [reply test/testRules.js:1556]: Added in fcaa5fb

As for code coverage - maybe we should consider changing our testing libraries to something newer? Nodeunit has been deprecated 5 years ago 😄 Jest has builtin coverage support

**Review by ehoogerbeets — APPROVED** [review]

**Pull request merged**